### PR TITLE
fix(completion): recursive `super::` completion in crate root

### LINKED ISSUE #12439: `use` erroneously suggests `super::` as a completion option even when it's not valid.
Consider you're in the `main.rs` of a crate and type:

```rust
use $0
```

Right now, this will offer `super::` as a completion, when there is *no* super. 

Additionally, following will offer a `super::` completion even when not relevant:

```rust
use super::$0;
use super::super::$0;
use super::super::super::$0;
// ... ad-infinitum...
```

---

(as an aside, I'm mainly opening these in response to the zulip thread asking for "minor inconveniences" in r-a's completion, hope y'all don't mind the issue spam!)

### PR DESCRIPTION
**Pull request by feniljain**
- should close #12439
- Suggest super according to the depth in tree

### CODE REVIEW

**Review by Veykril — COMMENTED** [review]
This doesn't only apply use paths, but to all paths, so we should add the `is_crate_root` check for `add_nameref_keywords_with_colon` to all of them, or rather, we should make `is_crate_root` a propery of `CompletionContext` and then just check for that inside `add_nameref_keywords_with_colon` without passing an extra bool

File: crates/ide-completion/src/tests/use_tree.rs [hunk 1]
```diff
@@ -273,11 +272,21 @@ pub use $0;
...
     );
 }
 
+#[test]
+fn pub_use_tree_super_in_crate_root() {
+    //https://github.com/rust-lang/rust-analyzer/issues/12439
+    check(
+        r#"
+pub use super::$0;
+"#,
```
Comment (Veykril) [inline crates/ide-completion/src/tests/use_tree.rs:285]: It makes more sense to test that we don't suggest a `super` that jumps out of a crate. The test case you've written here is already tested plenty now as can be seen by the diff in the test fixtures.

So the following test would test my other review part:
```suggestion
        r#"
mod foo {
    pub use super::$0;
}
"#,
```

File: crates/ide-completion/src/completions.rs [hunk 2]
```diff
@@ -103,11 +103,29 @@ impl Completions {
...
+
+        if ctx.depth_from_crate_root > 0 {
+            self.add_keyword(ctx, "super");
+        }
+    }
+
+    pub(crate) fn add_super_kw_acc_to_mod_depth(
+        &mut self,
+        super_chain_len: &usize,
+        ctx: &CompletionContext,
```
Comment (Veykril) [inline crates/ide-completion/src/completions.rs:124]:
```suggestion
        ctx: &CompletionContext,
        super_chain_len: usize,
```
order, and no need for the reference

File: crates/ide-completion/src/completions.rs [hunk 3]
```diff
@@ -103,11 +103,29 @@ impl Completions {
...
     pub(crate) fn add_nameref_keywords(&mut self, ctx: &CompletionContext) {
-        ["self", "super", "crate"].into_iter().for_each(|kw| self.add_keyword(ctx, kw));
+        ["self", "crate"].into_iter().for_each(|kw| self.add_keyword(ctx, kw));
+
+        if ctx.depth_from_crate_root > 0 {
+            self.add_keyword(ctx, "super");
+        }
+    }
+
+    pub(crate) fn add_super_kw_acc_to_mod_depth(
```
Comment (Veykril) [inline crates/ide-completion/src/completions.rs:121]:
```suggestion
    pub(crate) fn add_super_keyword(
```

File: crates/ide-completion/src/context.rs [hunk 4]
```diff
@@ -490,6 +496,20 @@ impl<'a> CompletionContext<'a> {
...
+            match mod_.parent(db) {
+                Some(mod_parent) => {
+                    mod_ = mod_parent;
+                    depth += 1;
+                }
+                None => {
+                    break;
+                }
+            }
+        }
```
Comment (Veykril) [inline crates/ide-completion/src/context.rs:511]:
```rs
let depth = iter::successors(mod_.parent(db), |it| it.parent(db)).count();
```
should be the same

File: crates/ide-completion/src/tests/use_tree.rs [hunk 5]
```diff
@@ -273,9 +271,39 @@ pub use $0;
             md bar
             kw crate::
             kw self::
+        "#]],
+    );
+}
+
+#[test]
+fn pub_suggest_use_tree_super_acc_to_depth_in_tree() {
```
Comment (Veykril) [inline crates/ide-completion/src/tests/use_tree.rs:279]: Let's add a third case here which should also show `super::`
```rs
mod foo {
    use $0;
}
```

File: crates/ide-completion/src/context.rs [hunk 6]
```diff
@@ -1285,8 +1306,9 @@ impl<'a> CompletionContext<'a> {
                 .map(|it| it.parent_path());
             path_ctx.qualifier = path.map(|path| {
                 let res = sema.resolve_path(&path);
-                let is_super_chain = iter::successors(Some(path.clone()), |p| p.qualifier())
-                    .all(|p| p.segment().and_then(|s| s.super_token()).is_some());
+                let super_chain_len = iter::successors(Some(path.clone()), |p| p.qualifier())
+                    .filter_map(|p| p.segment().and_then(|s| s.super_token()))
+                    .count();
```
Comment (Veykril) [inline crates/ide-completion/src/context.rs:1311]: So we only want to complete `super` when the entire path consists of `super` segments, your changes here just don't count non-super path which isn't quite correct, I think `super_chain_len` needs to be `Option<usize>`, None for when the path consists of something other than super segments, `Some` if its only super segments. An empty path would be `Some(0)`

**Comment by Veykril** [comment]
I've done some major refactorings to the ide-completion crate, so you might wanna manually port your changes over again instead of rebasing, sorry 😅

**Comment by feniljain** [comment]
Hey @Veykril , sorry for the delay. I have been stuck with something for a while.

Thanks for all the reviews. I have applied all of them and as mentioned I have manually ported all the changes and opened a new PR here: https://github.com/rust-lang/rust-analyzer/pull/12735

File: crates/ide-completion/src/tests/use_tree.rs (hunk 5 above)
Comment (feniljain) [reply crates/ide-completion/src/tests/use_tree.rs:279]: Added this test, alongwith one more for not suggesting `super::` when a non `super` kw has appeared in the segment chain

File: crates/ide-completion/src/context.rs (hunk 6 above)
Comment (feniljain) [reply crates/ide-completion/src/context.rs:1311]: That's a nice catch, thanks for pointing it out!

Now resolved

**Pull request closed without merging**